### PR TITLE
WIP: fast TPC space charge correction: shrink corrections outside of the TPC row

### PR DESCRIPTION
**Pull request by sgorbuno**
- shrink corrections that are outside of the TPC row to the values at the row boundary 
( via TPCFastSpaceChargeCorrection::setInterpolationSafetyMargin(float val); )

- when set from an analytical global correction: ignore the input data in gaps between the sectors ( outside of the readout area )

- multi-threaded creation of the SP correction 
( via TPCFastSpaceChargeCorrectionHelper::setNthreads(int n) or TPCFastSpaceChargeCorrectionHelper::setNthreadsToMaximum() )

- some code cleanup

### CODE REVIEW

File: GPU/TPCFastTransformation/TPCFastTransformGeo.h [hunk 1]
```diff
@@ -47,6 +47,9 @@ class TPCFastTransformGeo
     float u0;         ///< min. u coordinate
     float scaleUtoSU; ///< scale for su (scaled u ) coordinate
     float scaleSUtoU; ///< scale for u coordinate
+
+    /// get width in U
+    float getUwidth() const { return -2.f * u0; }
```
Comment (davidrohr) [inline GPU/TPCFastTransformation/TPCFastTransformGeo.h]: Please prepend by `GPUdi() ` to fix the GPU compilation

**Comment by davidrohr** [comment]
@sgorbuno : Why is this WIP? Can it be merged now that it passes the CI, or is more work needed?

**Comment by sgorbuno** [comment]
> @sgorbuno : Why is this WIP? Can it be merged now that it passes the CI, or is more work needed?

Hi David! 

Alex and Matthias want to verify if this PR fixes all the problems.

If they are ok with the PR, we can merge.

**Comment by davidrohr** [comment]
ok, thx. I will check later if it fixes the FPEs for me.
@sgorbuno : Can you confirm that the format of the maps did not change, i.e. I can reuse the existing maps? It just adds the "bounds-check", and improved map creation.

File: GPU/TPCFastTransformation/Spline2DHelper.cxx [hunk 2]
```diff
@@ -365,36 +365,38 @@ void Spline2DHelper<DataT>::approximateDataPoints(
...
-        int ind[17];
-        getScoefficients(iu, iv, fGridU.getKnot(pu).u, fGridV.getKnot(pv).u, c, ind);
-        c[16] = -1.;
-        ind[16] = ip;
-        // S = sum c[i]*Par[ind[i]]
-        double w = 1.e-8;
-        for (int i = 0; i < 17; i++) {
-          for (int j = i; j < 17; j++) {
-            solver.A(ind[i], ind[j]) += w * c[i] * c[j];
+  // add extra smoothness in case some data is missing
```
Comment (davidrohr) [inline GPU/TPCFastTransformation/Spline2DHelper.cxx:368]: Just for my understanding, why is this removed?

**Comment by sgorbuno** [comment]
> ok, thx. I will check later if it fixes the FPEs for me. @sgorbuno : Can you confirm that the format of the maps did not change, i.e. I can reuse the existing maps? It just adds the "bounds-check", and improved map creation.

@davidrohr, 
TPCFastSpaceChargeCorrection class got a new member fkInterpolationSafetyMargin. So I'm afraid you need to regenerate your maps.. :(

Or I can make it a constexpr if you like?

File: Detectors/TPC/calibration/src/TPCFastSpaceChargeCorrectionHelper.cxx [hunk 3]
```diff
@@ -356,11 +374,13 @@ void TPCFastSpaceChargeCorrectionHelper::testGeometry(const TPCFastTransformGeo&
   }
 }
 
-std::unique_ptr<o2::gpu::TPCFastSpaceChargeCorrection> TPCFastSpaceChargeCorrectionHelper::createSpaceChargeCorrection(
+std::unique_ptr<o2::gpu::TPCFastSpaceChargeCorrection> TPCFastSpaceChargeCorrectionHelper::createFromTrackResiduals(
   const o2::tpc::TrackResiduals& trackResiduals, TTree* voxResTree)
```
Comment (wiechula) [inline Detectors/TPC/calibration/src/TPCFastSpaceChargeCorrectionHelper.cxx:378]: I propose to implement options to switch between smoothed and non-smoothed and also to be able to invert signs:

```suggestion
std::unique_ptr<o2::gpu::TPCFastSpaceChargeCorrection> TPCFastSpaceChargeCorrectionHelper::createFromTrackResiduals(
  const o2::tpc::TrackResiduals& trackResiduals, TTree* voxResTree, bool useSmoothed, bool invertSigns)
```

File: Detectors/TPC/calibration/include/TPCCalibration/TPCFastSpaceChargeCorrectionHelper.h [hunk 4]
```diff
@@ -59,40 +59,34 @@ class TPCFastSpaceChargeCorrectionHelper
...
-    std::function<void(int roc, int irow, double y, double z, double& dx, double& dy, double& dz)> correctionLocal);
-
-  /// creates TPCFastSpaceChargeCorrection object
-  std::unique_ptr<TPCFastSpaceChargeCorrection> create(const int nKnotsY = 10, const int nKnotsZ = 20);
+      correctionGlobal,
+    const int nKnotsY = 10, const int nKnotsZ = 20);
+
+  /// Create SpaceCharge correction out of the voxel tree
+  std::unique_ptr<o2::gpu::TPCFastSpaceChargeCorrection> createFromTrackResiduals(
+    const o2::tpc::TrackResiduals& trackResiduals, TTree* voxResTree);
```
Comment (wiechula) [inline Detectors/TPC/calibration/include/TPCCalibration/TPCFastSpaceChargeCorrectionHelper.h:89]:
```suggestion
  std::unique_ptr<o2::gpu::TPCFastSpaceChargeCorrection> createFromTrackResiduals(
    const o2::tpc::TrackResiduals& trackResiduals, TTree* voxResTree, bool useSmoothed = false, bool invertSigns = false);
```

File: Detectors/TPC/calibration/src/TPCFastSpaceChargeCorrectionHelper.cxx [hunk 5]
```diff
@@ -485,10 +502,10 @@ std::unique_ptr<o2::gpu::TPCFastSpaceChargeCorrection> TPCFastSpaceChargeCorrect
...
     double dz = x * trackResiduals.getDZ2X(z2xBin);
 
-    double correctionX = v->D[o2::tpc::TrackResiduals::ResX];
-    double correctionY = v->D[o2::tpc::TrackResiduals::ResY];
-    double correctionZ = v->D[o2::tpc::TrackResiduals::ResZ];
-    double correctionD = v->D[o2::tpc::TrackResiduals::ResD];
+    double correctionX = v->DS[o2::tpc::TrackResiduals::ResX];
+    double correctionY = v->DS[o2::tpc::TrackResiduals::ResY];
+    double correctionZ = v->DS[o2::tpc::TrackResiduals::ResZ];
+    double correctionD = v->DS[o2::tpc::TrackResiduals::ResD];
```
Comment (wiechula) [inline Detectors/TPC/calibration/src/TPCFastSpaceChargeCorrectionHelper.cxx:508]: `correctionD` is not used, I think ...

```suggestion
    double correctionX = useSmoothed ? v->DS[o2::tpc::TrackResiduals::ResX] : v->D[o2::tpc::TrackResiduals::ResX];
    double correctionY = useSmoothed ? v->DS[o2::tpc::TrackResiduals::ResY] : v->D[o2::tpc::TrackResiduals::ResY];
    double correctionZ = useSmoothed ? v->DS[o2::tpc::TrackResiduals::ResZ] : v->D[o2::tpc::TrackResiduals::ResZ];

    if (invertSigns) {
      correctionX *= -1.;
      correctionY *= -1.;
      correctionZ *= -1.;
    }
```

File: GPU/TPCFastTransformation/TPCFastSpaceChargeCorrection.h [hunk 6]
```diff
@@ -247,6 +254,8 @@ class TPCFastSpaceChargeCorrection : public FlatObject
   RowInfo* mRowInfoPtr;           //! (transient!!) pointer to RowInfo array inside the mFlatBufferPtr buffer
   SliceRowInfo* mSliceRowInfoPtr; //! (transient!!) pointer to SliceRowInfo array inside the mFlatBufferPtr
 
+  float fInterpolationSafetyMargin{0.1}; // 10% area around the TPC row. Outside of this area the interpolation returns the boundary values.
```
Comment (wiechula) [inline GPU/TPCFastTransformation/TPCFastSpaceChargeCorrection.h:257]: Can this be put as last data member? Also, there is no ClassDef, but we have it in the LinkDef: https://github.com/AliceO2Group/AliceO2/blob/dev/GPU/TPCFastTransformation/TPCFastTransformationLinkDef_O2.h#L67
also the SliceInfo https://github.com/AliceO2Group/AliceO2/blob/dev/GPU/TPCFastTransformation/TPCFastTransformationLinkDef_O2.h#L66
Is this intentional?

File: GPU/TPCFastTransformation/macro/TPCFastTransformInit.C [hunk 7]
```diff
@@ -84,7 +84,7 @@ void TPCFastTransformInit(const char* fileName = "debugVoxRes.root",
 
   o2::tpc::TPCFastSpaceChargeCorrectionHelper* corrHelper = o2::tpc::TPCFastSpaceChargeCorrectionHelper::instance();
 
-  auto corrPtr = corrHelper->createSpaceChargeCorrection(trackResiduals, voxResTree);
+  auto corrPtr = corrHelper->createFromTrackResiduals(trackResiduals, voxResTree);
```
Comment (wiechula) [inline GPU/TPCFastTransformation/macro/TPCFastTransformInit.C:87]: Also here I would add to the function call `, bool useSmoothed = false, bool invertSigns = false` and then
```suggest
auto corrPtr = corrHelper->createFromTrackResiduals(trackResiduals, voxResTree, useSmoothed, invertSigns);
```

Comment (wiechula) [reply GPU/TPCFastTransformation/macro/TPCFastTransformInit.C:87]: And below again:
```c++
    double correctionX = useSmoothed ? v->DS[o2::tpc::TrackResiduals::ResX] : v->D[o2::tpc::TrackResiduals::ResX];
    double correctionY = useSmoothed ? v->DS[o2::tpc::TrackResiduals::ResY] : v->D[o2::tpc::TrackResiduals::ResY];
    double correctionZ = useSmoothed ? v->DS[o2::tpc::TrackResiduals::ResZ] : v->D[o2::tpc::TrackResiduals::ResZ];

    if (invertSigns) {
      correctionX *= -1.;
      correctionY *= -1.;
      correctionZ *= -1.;
    }
```

**Review by wiechula — CHANGES_REQUESTED** [review]

**Comment by davidrohr** [comment]
this works for me so far. The comments affect only the creation of the map in the helpers, so can be done in a seperate PR.
Merging this.

File: Detectors/TPC/calibration/src/TPCFastSpaceChargeCorrectionHelper.cxx (hunk 3 above)
Comment (sgorbuno) [reply Detectors/TPC/calibration/src/TPCFastSpaceChargeCorrectionHelper.cxx:378]: Hi @wiechula,  
I have put your changes in another PR: https://github.com/AliceO2Group/AliceO2/pull/11731
since this one is already merged
